File: objective-c/Constrain/Constrain/ViewControllers/MainViewController.m
//
//  MainViewController.m
//  Constrain
//
//  Created by Alexandr Kirilov on 2019-06-25.
//  Copyright © 2019 none. All rights reserved.
//

#import "MainViewController.h"

@interface MainViewController ()



@end

@implementation MainViewController

-(instancetype) init {

    self = [super init];
    if (self) {
		self.view.backgroundColor = [UIColor whiteColor];
    }
    return self;
}

-(void) viewDidLoad {

    [super viewDidLoad];
	
    UInt8 oPadding = 30;
    UInt8 oElementHeight = 50;
	
	UIButton *oButtonExit = [self mButtonExit];
	[self.view addSubview:oButtonExit];
	[NSLayoutConstraint constraintWithItem:oButtonExit
		attribute:NSLayoutAttributeCenterX relatedBy:NSLayoutRelationEqual toItem:self.view
		attribute:NSLayoutAttributeCenterX multiplier:1.0 constant:0
	].active = YES;
	[NSLayoutConstraint constraintWithItem:oButtonExit
		attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:self.view
		attribute:NSLayoutAttributeBottom multiplier:1.0 constant:-oPadding
	].active = YES;
	[NSLayoutConstraint constraintWithItem:oButtonExit
		attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:self.view
		attribute:NSLayoutAttributeLeft multiplier:1.0 constant:oPadding
	].active = YES;
	[NSLayoutConstraint constraintWithItem:oButtonExit
		attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:self.view
		attribute:NSLayoutAttributeRight multiplier:1.0 constant:-oPadding
	].active = YES;
	[NSLayoutConstraint constraintWithItem:oButtonExit
		attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:oButtonExit
		attribute:NSLayoutAttributeBottom multiplier:1.0 constant:-oElementHeight
	].active = YES;
	
	UILabel *oLabelLanguage = [self mLabelLanguage];
	[self.view addSubview:oLabelLanguage];
	[NSLayoutConstraint constraintWithItem:oLabelLanguage
		attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self.view
		attribute:NSLayoutAttributeTop multiplier:1.0 constant:2.5 * oPadding
	].active = YES;
	[NSLayoutConstraint constraintWithItem:oLabelLanguage
		attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:self.view
		attribute:NSLayoutAttributeLeft multiplier:1.0 constant:oPadding
	].active = YES;
	[NSLayoutConstraint constraintWithItem:oLabelLanguage
		attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:self.view
		attribute:NSLayoutAttributeRight multiplier:1.0 constant:-oPadding
	].active = YES;
	[NSLayoutConstraint constraintWithItem:oLabelLanguage
		attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:oLabelLanguage
		attribute:NSLayoutAttributeTop multiplier:1.0 constant:oElementHeight
	].active = YES;
	
	UILabel *oLabelApplication = [self mLabelApplication];
	[self.view addSubview:oLabelApplication];
	[NSLayoutConstraint constraintWithItem:oLabelApplication
		attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:oLabelLanguage
		attribute:NSLayoutAttributeTop multiplier:1.0 constant:oPadding
	].active = YES;
	[NSLayoutConstraint constraintWithItem:oLabelApplication
		attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:self.view
		attribute:NSLayoutAttributeLeft multiplier:1.0 constant:oPadding
	].active = YES;
	[NSLayoutConstraint constraintWithItem:oLabelApplication
		attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:self.view
		attribute:NSLayoutAttributeRight multiplier:1.0 constant:-oPadding
	].active = YES;
	[NSLayoutConstraint constraintWithItem:oLabelApplication
		attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:oLabelApplication
		attribute:NSLayoutAttributeTop multiplier:1.0 constant:oElementHeight
	].active = YES;
}


#pragma mark UI Elements

-(UIButton *) mButtonExit {

    UIButton *oButtonExit = [self mCreateButton:@"Exit" inFrame:CGRectMake(0,0,0,0)];
	oButtonExit.backgroundColor = [UIColor
		colorWithRed: 0.0 / 255.0 green: 62.0 / 255.0
		blue: 175.0 / 255.0 alpha: 1
	];
	[oButtonExit setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
	[oButtonExit addTarget:self action:@selector(mActionButtonExit) forControlEvents:UIControlEventTouchUpInside];
	oButtonExit.translatesAutoresizingMaskIntoConstraints = NO;
	return oButtonExit;
}

-(UILabel *) mLabelLanguage {

	UILabel *oLabelLanguage = [self mCreateLabel:@"Objective-C Examples" inFrame:CGRectMake(100, 100, 200, 20)];
	[oLabelLanguage setTextColor:[UIColor blackColor]];
	[oLabelLanguage setTextAlignment:NSTextAlignmentCenter];
	oLabelLanguage.translatesAutoresizingMaskIntoConstraints = NO;
	return oLabelLanguage;
}

-(UILabel *) mLabelApplication {

	UILabel *oLabelApplication = [self mCreateLabel:@"Constrains Usage" inFrame:CGRectMake(100, 150, 250, 20)];
	[oLabelApplication setTextColor:[UIColor blackColor]];
	[oLabelApplication setTextAlignment:NSTextAlignmentCenter];
	oLabelApplication.translatesAutoresizingMaskIntoConstraints = NO;
	return oLabelApplication;
}


#pragma mark UI Element primitives

-(UIButton *) mCreateButton: (NSString *)inTitle inFrame:(CGRect)inFrame {

    UIButton *oButton = [[UIButton alloc] initWithFrame:inFrame];
    [oButton setTitle:inTitle forState:UIControlStateNormal];
    return oButton;
}

-(UILabel *) mCreateLabel: (NSString *)inText inFrame:(CGRect)inFrame {

	UILabel *oLabel = [[UILabel alloc] initWithFrame:inFrame];
	oLabel.text = inText;
	return oLabel;
}


#pragma mark Button Actions

-(void) mActionButtonExit {

	[Handler mDoExit];
}

@end

File: swift/Localisation/v1/Localisation/ViewControllers/MainViewController.swift
//
//  MainViewController.swift
//  Localisation
//
//  Created by Alexandr Kirilov on 28/02/2020.
//  Copyright © 2020 Alexandr Kirilov. All rights reserved.
//

import UIKit

class MainViewController: UIViewController {

	var pScreenWidth: CGFloat = 0;
	var pScreenHeight: CGFloat = 0;
	var pElementX: CGFloat = 0;
	var pElementWidth: CGFloat = 0;
	
    override func viewDidLoad() {
        
		super.viewDidLoad()
		view.backgroundColor = _COLOR_WHITE;
	}
	
	override func viewDidLayoutSubviews() {
		
		super.viewDidLayoutSubviews();
		
		pScreenWidth = view.bounds.width;
		pScreenHeight = view.bounds.height;
		pElementX = (pScreenWidth - (pScreenWidth * 0.8)) / 2;
		pElementWidth = pScreenWidth * 0.8;
		
		mButtonExit();
        view.addSubview(mLabel(inText: "Swift examples", inY: 0.1, inHeight: 0.05));
		view.addSubview(mLabel(inText: "Localisation", inY: 0.15, inHeight: 0.1));
	}
	
	// MARK: UI Elements
	

	
	private func mButtonExit() -> Void {

		let oButton: UIButton = UIButton(frame: mCreateFrame(inY: 0.87, inHeight: 0.07));
		oButton.backgroundColor = _COLOR_BLUE_DARK;
		oButton.setTitleColor(.white, for: .normal);
		oButton.setTitle("Exit", for: .normal);
		oButton.addTarget(self, action: #selector(mDoExit(_:)), for: .touchUpInside);
		view.addSubview(oButton);
	}
	
	private func mLabel(inText: String, inY: CGFloat, inHeight: CGFloat) -> UILabel {

		let oLabel: UILabel = UILabel(frame: mCreateFrame(inY: inY, inHeight: inHeight));
		oLabel.text = inText;
		oLabel.textAlignment = NSTextAlignment.center;
		return oLabel;
	}
	
	private func mCreateSquare(inSideSize: CGFloat, inY: CGFloat) -> CGRect {
	
		return CGRect.init(
			x: (pScreenWidth - inSideSize) / 2,
			y: pScreenHeight * inY,
			width: inSideSize,
			height: inSideSize
		);
	}
	
	private func mCreateFrame(inY: CGFloat, inHeight: CGFloat) -> CGRect {

		return CGRect.init(
			x: pElementX, y: pScreenHeight * inY,
			width: pElementWidth, height: pScreenHeight * inHeight
		);
	}
	
	// MARK: Button Actions
	
	@objc func mDoExit(_ inSender: Any) -> Void {

		MainHandler.mDoExit();
	}
}
